### ISSUE TITLE
fix(check): a suppression marker with a trailing rationale is silently inert

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#1168]
## Summary

A `bca: suppress(...)` marker with anything after the closing paren on the
same line is rejected as malformed, so the suppression does not apply — and
the natural way to write the rationale `AGENTS.md` asks for is exactly that.

Measured (limit `nargs=5`, two identical 8-argument functions):

```console
$ bca check --no-config --threshold nargs=5 --no-summary --no-remediation -p /tmp/x
warning: /tmp/x/a.rs:6: malformed bca suppression marker body 'bca: suppress(nargs) — threaded context, not a god-function'
/tmp/x/a.rs:5-8: many_prose: nargs = 8 (limit 5)
```

The function whose marker is `// bca: suppress(nargs)` alone is suppressed.
The one whose marker carries a trailing rationale is reported.

## Why it matters

`AGENTS.md` ("Responding to bca metric feedback") tells contributors to
suppress "with a reason":

> add an in-source marker with a one-line rationale rather than contorting
> the code: `// bca: suppress(<metrics>)`

Writing that rationale on the marker line is the obvious reading, and it
voids the marker. The warning does fire, but it lands in a gate's output
among `skipped N` / `filtered N` lines that are routine, so the practical
outcome is a suppression the author believes is active and is not.

This is the second instance of the same class: an unknown metric name also
warns *and* voids the whole marker. Both share the shape "the marker is
almost right, so it is silently inert".

## Options

1. **Accept trailing prose** after the metric list — treat the rest of the
   line as the rationale. This is what authors are already writing, and it
   puts the reason where a reader of the flagged function will see it.
2. Keep the strict grammar but make the warning impossible to miss (name the
   function, and exit non-zero on a malformed marker rather than continuing
   with it inert).

(1) is the better fit for the documented workflow. Either way `AGENTS.md`
should show the accepted spelling explicitly, since it currently shows a form
that does not work.

---

## Resolution Plan

**Take option 1: accept trailing prose after the metric list.**

The decisive argument is the one the Summary makes — `AGENTS.md` instructs
contributors to suppress "with a reason", shows `// bca: suppress(<metrics>)`,
and the obvious way to combine those two instructions is the spelling that
silently voids the marker. A grammar that rejects what the documentation asks
for is the grammar that is wrong, not the author.

### Why not option 2

Option 2 (keep the strict grammar, make the warning louder) has a specific
flaw: the failure mode is a suppression the author *believes is active*. Making
the warning louder helps the author who reads it at the moment of writing, and
does nothing for every later reader of that file, who sees a marker that looks
correct. The marker is the durable artifact; it should mean what it appears to
mean.

Exiting non-zero on a malformed marker — the second half of option 2 — is worth
keeping as a *supplement* to option 1, for the genuinely malformed remainder
(see step 4).

### The shared class matters more than the instance

The Summary notes this is the second instance of "the marker is almost right,
so it is silently inert" — the first being an unknown metric name, which warns
and voids the whole marker. That generalisation is the useful output of this
issue, and the fix should address the class:

**A marker that is recognisably a `bca: suppress` marker should never
silently do nothing.** Either it suppresses what it names, or the run fails.
"Warn and continue as if the marker were absent" is the state that produces
both instances.

For the unknown-metric case that means: suppress the metrics that *are*
recognised and warn about the rest, rather than voiding the marker wholesale.
`// bca: suppress(cognitive, exit)` should still suppress `cognitive` — losing
that is surprising, and `exit`-for-`nexits` is a documented-in-AGENTS.md
mistake people actually make.

### Steps

1. Extend the marker grammar: after the closing paren, accept an optional
   remainder on the same line and treat it as the rationale. Accept the common
   separators without requiring one (`—`, `-`, `:`, `//`, or bare prose), since
   the point is to stop being fussy.
2. Decide whether the rationale is captured or discarded. Recommend
   **capturing** it and surfacing it in `--report-suppressed` / the SARIF
   `suppressions` justification field — that turns the rationale from a comment
   into data, and it is a natural fit for a field that already exists.
3. Apply the class fix: an unrecognised metric name warns and is skipped, but
   the recognised names in the same marker still suppress.
4. Keep a genuine-malformation path: something that opens `bca: suppress` but
   has no parseable metric list still warns. Consider making that a hard error
   under `--no-suppress`-adjacent strictness, but **not** a hard error by
   default — a stray mention of the marker syntax inside a doc comment or a
   test fixture should not fail the gate.
5. Tests: the issue's exact reproducer (two identical 8-argument functions, one
   marker bare and one with a trailing rationale, both suppressed); each
   separator form; a mixed known/unknown metric list suppressing the known
   half; `suppress-file` with a rationale; and a genuinely malformed marker
   still warning.
6. **Fix `AGENTS.md` in the same commit.** It currently shows a form whose
   natural completion does not work; it should show the accepted spelling
   *with* a rationale, since that is the workflow it is prescribing. Same for
   `big-code-analysis-book/src/commands/suppression.md` and
   `recipes/agent-feedback.md`.
7. Sweep the repo for existing markers that carry trailing prose and are
   therefore inert today. Any that exist are functions currently being gated
   that their author believed were exempt — worth knowing before the fix
   changes their status, because they will appear as *newly suppressed* rather
   than as a fix.

## Assessment

| Dimension  | Rating |
|------------|--------|
| Difficulty | Low |
| Complexity | Medium |
| Priority   | High |

**Difficulty — Low.** A parser relaxation plus the class fix in step 3. The
test matrix is the bulk of it.

**Complexity — Medium.** It touches the suppression parser, the SARIF
suppressions rendering if step 2 is taken, three documentation surfaces, and
potentially this repo's own marker set (step 7) — and step 7 can move the
self-scan baseline.

**Priority — High.** The project documents a workflow that does not work, and
the failure is invisible in the direction that matters: the author believes the
suppression is active, the gate disagrees, and the warning that says so is
buried among routine `skipped N` / `filtered N` lines. `AGENTS.md`'s
anti-gaming guidance depends on suppression-with-a-reason being the easy path;
right now it is the broken one.

---

## Resolution

Fixed on `fix/issue-1168` (commit `9035491d`), taking option 1.

**Grammar.** Anything after the metric list is the rationale; no separator is required and none is privileged. After a *bare* verb a separator (`-`, `:`, `//`, `#`, em/en dash) **is** required — with no list there is nothing to distinguish a reason from a sentence about the marker, and `// bca: suppress markers are honoured here` must not silence a function.

**Class fix.** An unusable metric name is skipped and reported; the recognised names in the same marker still suppress. This reverses the void-on-unknown contract that #896 pinned, and that test now pins the new behaviour alongside a companion proving an all-unusable list still suppresses nothing. A body that parses to no directive at all (unknown verb, unbalanced paren, bare verb + prose) still warns, still non-fatally.

**Deferred to 3.0.** Step 2 — capturing the rationale into `--report-suppressed` / the SARIF `justification` — needs a new public field on `OffenderRecord` and on `SuppressionMarker`; neither is `#[non_exhaustive]`, so that is a SemVer-major break under `STABILITY.md`. The rationale is parsed and discarded for now. The fix nonetheless corrects the classification: a rationale-carrying marker's offender now renders as SARIF `suppressions[].kind: inSource` instead of an active finding.

**Step 7 sweep.** Zero inert markers in this repository (a full `bca check --no-config` scan over every source tree emitted no marker warnings pre-fix), so nothing becomes newly suppressed and the self-scan baseline did not move.

Docs updated in the same commit: `AGENTS.md`, the book's `commands/suppression.md` (new *Rationale text* section) and `recipes/agent-feedback.md`, and `.claude/hooks/bca-guidance.txt`. `make pre-commit` and `make chain-audit` are both green.

**Comment by dekobon** [comment]
Fixed on `fix/issue-1168` (commit `9035491d`).

**What changed**

- Anything after the metric list is now the rationale — no separator required, none privileged (`—`, `–`, `-`, `:`, `//`, bare prose all work). After a *bare* verb (`bca: suppress` with no list) a separator **is** required; see the correction below.
- The class fix: an unusable metric name is skipped and reported rather than voiding the marker. `bca: suppress(cognitive, exit)` silences `cognitive` and warns about `exit`. Skipping can only ever narrow what a marker covers, so a typo cannot widen scope.
- Genuine malformation (unknown verb, unbalanced paren, bare verb + prose) still warns and is still non-fatal. Its message now names the accepted shapes instead of only echoing the body.
- `parse_marker` returns a `MarkerScan { suppression, diagnostics }` instead of a `Result`, since a marker can now be partly usable. Crate-internal type; no published shape changed.

**Corrections to the issue**

1. *The reproducer as written does not reproduce.* Both markers sit above their functions, outside the body, so neither attaches — the bare one does not suppress either, and pre-fix both functions are reported. Moving the markers inside the bodies reproduces exactly as described. The reproducer in the commit's test (`a_trailing_rationale_suppresses_exactly_like_a_bare_marker`) uses the in-body form.
2. *Step 2 (capture the rationale, surface it in `--report-suppressed` / SARIF `justification`) is deferred.* Carrying the text to SARIF needs a new public field on `OffenderRecord`, and surfacing it in `bca exemptions` needs one on `SuppressionMarker`. Neither carries `#[non_exhaustive]`, so adding a field is a SemVer-major break per `STABILITY.md` ("existing items may gain fields *where they already permit it*") and AGENTS.md's "never slip a SemVer break into a patch or minor". The rationale is parsed and discarded for now; filed as a 3.0 follow-up.

   The fix does still improve `--report-suppressed`: a rationale-carrying marker now classifies its offender as *suppressed* (SARIF `suppressions[].kind: inSource`) where before it appeared as an active finding.
3. *`suppress(metric, reason = "...")` is no longer the plan.* The book reserved that syntax for a future release; the trailing-prose form supersedes it, and the reserved-syntax section now says so.
4. *Step 4's "hard error under a strictness flag" was not added.* No flag exists to hang it on, and the warning-only default is what keeps a doc comment or test fixture mentioning the syntax from failing a gate.

**Step 7 sweep — no inert markers in this repo**

Every inert marker (trailing prose *or* unusable metric name) emits a walk-time `warning:`. A full pre-fix scan —

```console
$ bca check --no-config --threshold cognitive=100000 \
    -p src -p big-code-analysis-cli -p big-code-analysis-web \
    -p big-code-analysis-py -p tests -p xtask -p big-code-analysis-bench
```

— produced zero warnings, so no marker here was silently inert and nothing becomes newly suppressed. `make self-scan` and `make self-scan-headroom` both pass unchanged; **the baseline did not move**.

**Validation**

`make pre-commit` green (`MAKE_EXIT=0`), `make chain-audit` green. Each new assertion was verified by perturbing the production line it names — requiring a separator after `)`, accepting any bare-verb trailer, restoring void-on-unknown, and swallowing the diagnostics — and each perturbation failed the intended tests and no others outside that invariant.

**Docs updated in the same commit:** `AGENTS.md`, `big-code-analysis-book/src/commands/suppression.md` (new *Rationale text* section), `recipes/agent-feedback.md`, `.claude/hooks/bca-guidance.txt`.

**Comment by dekobon** [comment]
Follow-up for the deferred rationale capture (step 2): #1175.
